### ISSUE TITLE
Refactor bam_to_h5.R

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#61]
Summary of TODOs for this issue (derived from comments below):

- [x] Remove `ggplot2` and `tidyr` dependencies.
- [x] Remove `is_test_run` as `test_bam_to_h5.R` (#238) can now be used.
- [x] Remove commented-out test code.
- [x] Conform to R styles, see `docs/developer/coding-style.md`.
- [x] Add Roxygen2-compliant comments.
- [x] Provide complementary documentation.
- [x] Specify feature as a parameter e.g. `CDS`, `ORF`, `uORF`.
- [x] Save H5 attributes `reads_total`, `buffer_left`, `reads_by_len` as `H5T_STD_I32LE`, not `H5T_IEEE_F64LE`.
- [x] Take all CDS-related co-ordinates from the CDS, not from UTR3 or UTR5.

Where appropriate, extend tests in `rscripts/tests/testthat/test_bam_to_h5.R`.

**Comment by ewallace** [comment]
* ggplot2 and tidyr dependencies not needed, remove.
* ontological confusion about gene/ORF/CDS needs fixing. (this is functionality more than refactoring, but noting here)
* ideally would get "read counts by position and length" for features with type as an argument, for example "CDS", or "ORF", or able to include "uORF"
* all CDS-related co-ordinates should be taken from the CDS *not* from UTR3 or UTR5.
* test data should come from #112

**Comment by FlicAnderson** [comment]
Just tagging in [this "memory not mapped" error I got with a dataset](https://github.com/riboviz/example-datasets/pull/17#issuecomment-707098000) because it's maybe worth referencing in documentation about `bam_to_h5.R`? It's possibly something for another issue, but since documentation will have to be updated following this refactoring, I'm pinning it here in case it's a helpful reminder. 

Also reminder about the importance of testing (briefly mentioned in #62, and following #213 ) and perhaps whether the testing for `bam_to_h5.R`should be split off as a separate issue during the refactoring process to allow this issue to be more specific.

**Comment by mikej888** [comment]
Requires #238.

**Comment by mikej888** [comment]
For buffer, either command-line OR GFF value is used:

* Find out which.
* Edward wants this only to be taken from the GFF and update code and tests accordingly.
* #83 highlights the ambiguity of "buffer" in different places in the workflow.
 
Check count matrix is an integer (see #69)

Add Roxygen2-compliant comments - see Flic's code.

**Comment by mikej888** [comment]
In relation to "Check count matrix is an integer #69:

| R variable/value | class | typeof | H5 variable | DATATYPE |
| - | ----- | ------ | -- | -------- |
| sum(output) | numeric | double | reads_total | H5T_IEEE_F64LE |
| start_codon_loc-1 | numeric | double | buffer_left | H5T_IEEE_F64LE |
| ncol(output)-stop_cod[3] | integer | integer | buffer_right | H5T_STD_I32LE |
| start_cod | integer | integer | start_codon_pos | H5T_STD_I32LE |
| stop_cod | integer | integer | stop_codon_pos | H5T_STD_I32LE |
| read_range | integer | integer | lengths | H5T_STD_I32LE |
| apply(output,1,sum) | numeric | double | reads_by_len | H5T_IEEE_F64LE |
| output | matrix | double | data | H5T_STD_I32LE |

Note that even though all attributes values (all but `data`) are written using `h5writeAttribute.integer`, the types in the H5 file correspond to the R types.

**Comment by mikej888** [comment]
In relation to Check count matrix is an integer #69:

| R variable/value | class | typeof | H5 variable | DATATYPE |
| - | ----- | ------ | -- | -------- |
| sum(output) | numeric | double | reads_total | H5T_IEEE_F64LE |
| start_codon_loc-1 | numeric | double | buffer_left | H5T_IEEE_F64LE |
| ncol(output)-stop_cod[3] | integer | integer | buffer_right | H5T_STD_I32LE |
| start_cod | integer | integer | start_codon_pos | H5T_STD_I32LE |
| stop_cod | integer | integer | stop_codon_pos | H5T_STD_I32LE |
| read_range | integer | integer | lengths | H5T_STD_I32LE |
| apply(output,1,sum) | numeric | double | reads_by_len | H5T_IEEE_F64LE |
| output | matrix | double | data | H5T_STD_I32LE |

Note that even though all attributes values (all but `data`) are written using `h5writeAttribute.integer`, the types in the H5 file correspond to the R types.

Note that for `reads_total` and `reads_by_len` to be written as integers requires either the values computed for these attributes to be cast to `integer` or for `output` to be created as a matrix of `integer`, not `double`.

One solution is to cast the results to integer, using `as.integer`, prior to saving using `h5.writeAttribute.integer`. For example:

```
h5writeAttribute.integer(as.integer(sum(output)),gid,name="reads_total")
h5writeAttribute.integer(as.integer(start_codon_loc-1),gid,name="buffer_left")
h5writeAttribute.integer(as.integer(apply(output,1,sum)),gid,name="reads_by\_len")
```

An alternative approach for, `sum(output) and `apply(output,1,sum)` is to ensure the `output` matrix is a matrix of `integer`. According to https://stackoverflow.com/questions/36943953/fastest-way-to-coerce-matrix-to-integer-matrix-in-r, a straightforward way to do the latter is:

```R
mode(output) <- "integer"
```

A quick test of this gives, in the H5 file, `reads_total` and `reads_by_len` attributes with `DATATYPE` `H5T_STD_I32LE` as desired.

**Comment by mikej888** [comment]
[refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61)

**Comment by mikej888** [comment]
Commit 0082bd3 of [refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61) completes:

* Provide complementary documentation.

**Comment by mikej888** [comment]
> Just tagging in [this "memory not mapped" error I got with a dataset](https://github.com/riboviz/example-datasets/pull/17#issuecomment-707098000) because it's maybe worth referencing in documentation about `bam_to_h5.R`? It's possibly something for another issue, but since documentation will have to be updated following this refactoring, I'm pinning it here in case it's a helpful reminder.
> 
> Also reminder about the importance of testing (briefly mentioned in #62, and following #213 ) and perhaps whether the testing for `bam_to_h5.R`should be split off as a separate issue during the refactoring process to allow this issue to be more specific.

Raised Investigate bam_to_h5.R 'memory not mapped' error #301

**Comment by mikej888** [comment]
Commit 84cf1a1 of [refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61):

* Implemented Fix bam_to_h5.R bug if is_riboviz_gff is FALSE #281.
* Updated to specify feature as a parameter
  - bam_to_h5.R now takes optional `--feature` parameter (default `CDS`).
  - References to 'CDS' replaced by references to 'feature' in `bam_to_h5.R`, `test_bam_to_h5.R`, `docs/reference/hdf5-data.md`.
  - `docs/reference/hdf5-data.md` documents that default used is currently `CDS`.
  - Created `data/Mok-tinysim-gffsam/tiny_2genes_20utrs_orf.gff3` a copy of `data/Mok-tinysim-gffsam/tiny_2genes_20utrs.gff3` with `CDS` replaced by `ORF`, for testing, and updated `docs/reference/data.md` to note this.
  - Added test with `--feature=ORF` to `test_bam_to_h5.R`
* Updated to ensure all HDF5 attributes and values are `H5T_STD_I32LE` and not `H5T_IEEE_F64LE` (specifically, `buffer_left`, `reads_by_len`, `reads_total`).

Implications of update to ensure all HDF5 attributes and values are `H5T_STD_I32LE` and not `H5T_IEEE_F64LE` on regression tests to be checked. A new regression test data set may be required since files will differ in recorded `DATATYPE` for attributes `buffer_left`, `reads_by_len`, `reads_total`.

Implications of specify feature as a parameter e.g. CDS, ORF, uORF to be discussed:

* Rename command-line parameter from `stop_in_cds` to `stop_in_feature`?
* Rename YAML configuration value from `stop_in_cds` to `stop_in_feature`?
* Add new `feature` configuration parameter to YAML?

**Comment by mikej888** [comment]
Exploring...

> Implications of update to ensure all HDF5 attributes and values are `H5T_STD_I32LE` and not `H5T_IEEE_F64LE` on regression tests to be checked. A new regression test data set may be required since files will differ in recorded `DATATYPE` for attributes `buffer_left`, `reads_by_len`, `reads_total`.

Run workflow and tests:
```console
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml
$ pytest -vs riboviz/test/regression/test_regression.py::test_bam_to_h5_h5 \
  --expected=$HOME/regression-test-data-20210119/ \
  --config-file=vignette/vignette_config.yaml \
  --nextflow --skip-workflow
...
riboviz/test/regression/test_regression.py::test_bam_to_h5_h5[vignette/output-WTnone] PASSED
riboviz/test/regression/test_regression.py::test_bam_to_h5_h5[vignette/output-WT3AT] PASSED
...
========================= 2 passed, 1 warning in 0.78s =========================
```
They pass! Check the file contents do indeed differ:
```console
$ h5dump vignette/output/WTnone/WTnone.h5 > WTnone.actual.h5.txt
$ h5dump ~/regression-test-data-20210119/output/WTnone/WTnone.h5 > WTnone.expected.h5.txt 
$ diff WTnone.expected.h5.txt WTnone.actual.h5.txt  | more
1c1
< HDF5 "/home/ubuntu/regression-test-data-20210119/output/WTnone/WTnone.h5" {
---
> HDF5 "vignette/output/WTnone/WTnone.h5" {
7c7
<                DATATYPE  H5T_IEEE_F64LE
---
>                DATATYPE  H5T_STD_I32LE
30c30
<                DATATYPE  H5T_IEEE_F64LE
---
>                DATATYPE  H5T_STD_I32LE
39c39
<                DATATYPE  H5T_IEEE_F64LE
---
>                DATATYPE  H5T_STD_I32LE
12021c12021
<                DATATYPE  H5T_IEEE_F64LE
---
...
```
Check using `h5diff`, which is used by the regression tests:
```console
$ h5diff vignette/output/WTnone/WTnone.h5 ~/regression-test-data-20210119/output/WTnone/WTnone.h5
--------------------------------
Some objects are not comparable
--------------------------------
Use -c for a list of objects.
$ echo $?
0
```
Exit code is 0 which is why the tests pass. Use `-c`:
```console
$ h5diff -c vignette/output/WTnone/WTnone.h5 ~/regression-test-data-20210119/output/WTnone/WTnone.h5
Not comparable: <buffer_left> is of class H5T_INTEGER and <buffer_left> is of class H5T_FLOAT
Not comparable: <reads_by_len> is of class H5T_INTEGER and <reads_by_len> is of class H5T_FLOAT
Not comparable: <reads_total> is of class H5T_INTEGER and <reads_total> is of class H5T_FLOAT
```
This is what was expected.

Check the exit code if `h5diff` is given two different H5 files, with different data:
```
$ h5diff vignette/output/WTnone/WTnone.h5 ~/regression-test-data-20210119/output/WT3AT/WT3AT.h5 
dataset: </YAL001C/vignette/reads/data> and </YAL001C/vignette/reads/data>
22 differences found
dataset: </YAL002W/vignette/reads/data> and </YAL002W/vignette/reads/data>
15 differences found
dataset: </YAL003W/vignette/reads/data> and </YAL003W/vignette/reads/data>
711 differences found
...
$ echo $?
1
```
OK so `h5diff` returns a non-zero exit code if `DATASET`s differ but not `ATTRIBUTE`s.

This is not a concern. The unit tests validate that `bam_to_h5.R` produces a valid H5 file with the expected values for `buffer_left`, `reads_by_len`, and `reads_total`.

The regression tests will catch any differences in the actual `DATASET`s in the H5 file. The regression test data will updated anyway as part of the next release.

**Comment by mikej888** [comment]
For Take all CDS-related co-ordinates from the CDS, not from UTR3 or UTR5, commit 5ba083e of [refactor-bam-to-h5-61-coords-from-cds](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61-coords-from-cds) (branched from [refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61)) implements what I think is the solution to this. 

* Both `start_codon_loc` and `stop_codon_loc` are deduced from the `gene_location` information for the feature, from the GFF.
* If `is_riboviz_gff` and `stop_in_cds` are both false the the stop codon location is taken to be 1 nucleotide beyond the end of the feature.

See the commit for details of the changes.

**Comment by mikej888** [comment]
Fixed Replace NULL with NA in bam_to_h5.R make_option calls #78

* `bam_to_h5.R` `--secondary-id` option has default value `NA`. Checks for `is.null` replaced with `is.na`.
* `test_bam_to_h5.R` function `RunSamToBamToH5` constucts `bam_to_h5.R` invocation to include `--secondary-id` if `secondary_id` is not `NA`.
* `prep_riboviz.nf|.py` only add `--secondary-id=<secondary_id>` to `bam_to_h5.R` invocation if configuration defines `secondary_id` and it is not `null|NULL`.

**Comment by mikej888** [comment]
@ewallace please comment on...

For Take all CDS-related co-ordinates from the CDS, not from UTR3 or UTR5, commit 5ba083e of [refactor-bam-to-h5-61-coords-from-cds](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61-coords-from-cds) (branched from [refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61)) implements what I think is the solution to this:

* Both `start_codon_loc` and `stop_codon_loc` are deduced from the `gene_location` information for the feature, from the GFF.
* If `is_riboviz_gff` and `stop_in_cds` are both false the the stop codon location is taken to be 1 nucleotide beyond the end of the feature.
* See the commit message for details of the changes.

Implications of specify feature as a parameter e.g. CDS, ORF, uORF:

* You commented 'ideally would get "read counts by position and length" for features with type as an argument, for example "CDS", or "ORF", or able to include "uORF"'. What does 'able to include "uORF"' mean?
* `bam_to_h5.R` now supports an optional `--feature` parameter, default `CDS`, which is used to specify the feature. There is a copy of `data/Mok-tinysim-gffsam/tiny_2genes_20utrs.gff3` called `data/Mok-tinysim-gffsam/tiny_2genes_20utrs_orf.gff3` which has `ORF` instead of `CDS` which is used by `test_bam_to_h5.R` to test this.
* There is a need to:
   - Rename command-line parameter from `stop_in_cds` to `stop_in_feature`.
   - Rename YAML configuration value from `stop_in_cds` to `stop_in_feature`.
   - As these break backward compatibility an alternative is to add `stop_in_feature` as a new parameter and, if both are provided, use `stop_in_feature` in preference.
   - Add new `feature` configuration parameter (default `CDS`).
   - Once decided, these can be subject of new issues.

Fix ontological confusion about gene/ORF/CDS (this is functionality more than refactoring, but noting here).

* Does this incur any work w.r.t this specific ticket?

**Comment by mikej888** [comment]
Discussed "Take all CDS-related co-ordinates from the CDS, not from UTR3 or UTR5, commit 5ba083e of refactor-bam-to-h5-61-coords-from-cds (branched from refactor-bam-to-h5-61) implements what I think is the solution to this" with Edward today and agreed it was fine.

**Comment by mikej888** [comment]
For Implications of specify feature as a parameter e.g. CDS, ORF, uORF:

* [refactor-bam-to-h5-feature-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-feature-61) branch off of [refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61) has added `feature` (default `CDS`) and `stop_in_feature` (default `FALSE`) parameters.
* `stop_in_feature` replaces `stop_in_cds` which is now deprecated. If both `stop_in_feature` and `stop_in_cds` are defined in a configuration then `stop_in_feature` takes precedence.
* Once this is merged create a new issue: Name: Remove `stop_in_cds` parameter. Description: Remove all occurrences from: docs/user/prep-riboviz-config.md, docs/reference/hdf5-data.md, prep_riboviz.nf, riboviz/workflow.py. Retain occurrences in `riboviz/params.py` and `riboviz/upgrade_config.py` so `riboviz.tools.upgrade_config_file` still works.

Fix ontological confusion about gene/ORF/CDS (this is functionality more than refactoring, but noting here).

* Does this incur any work w.r.t this specific ticket?

**Comment by mikej888** [comment]
Merged [refactor-bam-to-h5-feature-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-feature-61) branch into [refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61).

**Comment by mikej888** [comment]
Commit d5fe636 of [refactor-bam-to-h5-61](https://github.com/riboviz/riboviz/tree/refactor-bam-to-h5-61):

bam_to_h5_functions.R checks that at least one occurrence of feature is in GFF. Previously the error would be:
```
Error: subscript contains out-of-bounds indices
Execution halted
```
Now it is, for example:
```
No XXX features found in /home/ubuntu/riboviz/data/Mok-tinysim-gffsam/tiny_2genes_20utrs.gff3
Execution halted
```

Extracted duplicated code from test_bam_to_h5.R and test_bam_to_h5_functions.R into bam_to_h5_helpers.R

**Comment by mikej888** [comment]
Once pull request #315 is merged, create new issue to remove now-deprecated `stop_in_cds` parameter at a future release.

* Remove stop_in_cds parameter.
* Description: Remove all occurrences from: docs/user/prep-riboviz-config.md, docs/reference/hdf5-data.md, prep_riboviz.nf, riboviz/workflow.py. Retain occurrences in riboviz/params.py and riboviz/upgrade_config.py so riboviz.tools.upgrade_config_file still works.
